### PR TITLE
Added S3 storage capability

### PR DESCRIPTION
**Pull request by jhunzik**

### CODE REVIEW

File: ingest/src/main/resources/application.yaml [hunk 1]
```diff
@@ -0,0 +1,6 @@
+s3Properties:
```
Comment (jhunzik) [inline ingest/src/main/resources/application.yaml]: This file isn't used and will be removed

File: ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+@Service
+public class S3Adaptor implements Adaptor {
+
+  private S3Client s3Client;
+
+  // We'll probably change these when we move to a cloud friendly configuration
+  private static final String S3_BUCKET = "staged-products";
+  private static final String S3_ENDPOINT = "http://127.0.0.1:9000";
```
Comment (vinamartin) [inline ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: If you keep the properties file, you should be able to do something like this in Spring:
```
@Value("${s3.endpoint}")
private string S3endpoint;
```

To populate it from the config file.

Comment (jhunzik) [reply ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: I had that originally and I can't remember why I changed it. I think we may be able to override those properties with env variables or similar when running in docker.

Comment (oconnormi) [reply ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: yes if you have the `@Value("${s3.endpoint}")` in your Configuration you can override it via `application.yml`, `ENV` or `--s3.endpoint=<stuff>`

**Comment by oconnormi** [comment]
I think you will need to include a configuration property to allow pointing at a file containing an aws access key that can be read at runtime so that this will work with docker swarm secrets

File: ingest/build.gradle [hunk 3]
```diff
@@ -13,5 +13,11 @@ docker {
 	buildArgs([JAR_FILE: "${tasks.bootJar.outputs.files.singleFile.name}"])
 }
 
+dependencies {
+	implementation platform("software.amazon.awssdk:bom:2.5.29")
+	implementation "software.amazon.awssdk:s3"
+
```
Comment (Kjames5269) [inline ingest/build.gradle]: What's the difference between `platform("x")` and `x`?

File: ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Value("${aws.s3.endpointUrl}")
+  private String S3_ENDPOINT;
+
+  @PostConstruct
+  private void initializeS3() {
+    s3Client =
+        S3Client.builder()
+            .endpointOverride(URI.create(S3_ENDPOINT))
+            .region(Region.US_GOV_EAST_1)
```
Comment (Kjames5269) [inline ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: ❓ should this be customizable?

File: ingest/src/main/java/com/connexta/ingest/service/api/IngestRequest.java [hunk 5]
```diff
@@ -6,9 +6,24 @@
...
  **
- * Java object representation of the json request
+ * Java object representation of the ingest request
  *
  */
-public interface IngestRequest {}
+@Getter
+@Setter
+@AllArgsConstructor
+public class IngestRequest {
```
Comment (Kjames5269) [inline ingest/src/main/java/com/connexta/ingest/service/api/IngestRequest.java]: This class can be found in the Ion-Ingest-API in the `InlineObject` class. Should we be using that class instead after the Owasp errors get fixed?

File: ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java (hunk 4 above)
Comment (vinamartin) [reply ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: ❗️ I would think so

File: ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+import software.amazon.awssdk.services.s3.model.S3Exception;
+
+@Service
+public class S3Adaptor implements Adaptor {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(S3Adaptor.class);
+  private S3Client s3Client;
+
+  @Autowired private String s3BucketQuarantine;
+
```
Comment (vinamartin) [inline ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: I think typically using `@Autowired` on the constructor and using setters is safer than inlining it here.

File: ingest/src/main/java/com/connexta/ingest/service/api/IngestRequest.java [hunk 7]
```diff
@@ -6,9 +6,24 @@
...
 /*
  **
- * Java object representation of the json request
+ * Java object representation of the ingest request
  *
  */
-public interface IngestRequest {}
+@Getter
+@Setter
+@AllArgsConstructor
```
[the remark this answers is not in the document]
Comment (jhunzik) [reply ingest/src/main/java/com/connexta/ingest/service/api/IngestRequest.java]: We don't need the setter, good catch.

File: ingest/src/main/java/com/connexta/ingest/service/api/IngestRequest.java (hunk 5 above)
Comment (jhunzik) [reply ingest/src/main/java/com/connexta/ingest/service/api/IngestRequest.java]: I planned on replacing it once we merged your other PR.

File: ingest/build.gradle (hunk 3 above)
Comment (jhunzik) [reply ingest/build.gradle]: Maven's BOM(Bill of Materials) is a kind of platform that is used to keep the sdk versions up to date.

File: ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java (hunk 2 above)
Comment (jhunzik) [reply ingest/src/main/java/com/connexta/ingest/adaptors/S3Adaptor.java]: I moved the S3 configuration over to the Docker side of things.

File: docker-compose.yml [hunk 8]
```diff
@@ -36,6 +36,17 @@ services:
...
         condition: any
+    secrets:
+      - source: aws_s3_secret
+        target: /secrets/aws_s3_secret.sec
+      - source: aws_s3_access
+        target: /secrets/aws_s3_access.sec
+    command:
+      - "--aws.s3.secret.file=/secrets/aws_s3_secret.sec"
+      - "--aws.s3.access.file=/secrets/aws_s3_access.sec"
+      - "--aws.s3.endpointUrl=http://s3.us-gov-west-1.amazonaws.com/"
```
Comment (emmberk) [inline docker-compose.yml]: ❗️remove this

File: docker-compose.yml [hunk 9]
```diff
@@ -36,6 +36,17 @@ services:
...
+      - "--aws.s3.bucket.quarantine=ingest-quarantine"
+      - "--aws.s3.region=us-gov-west-1"
```
Comment (emmberk) [inline docker-compose.yml]: ❗️remove this

File: ingest/build.gradle [hunk 10]
```diff
@@ -27,5 +27,11 @@ docker {
 	buildArgs([JAR_FILE: "${tasks.bootJar.outputs.files.singleFile.name}"])
 }
 
+dependencies {
+	implementation platform("software.amazon.awssdk:bom:2.5.29")
+	implementation "software.amazon.awssdk:s3"
+	compile group: "commons-io", name: "commons-io", version: "2.6"
```
Comment (jhunzik) [inline ingest/build.gradle]: I'll remove this as it's not used anymore.

**Comment by emmberk** [comment]
build now

**Comment by emmberk** [comment]
The CI build failed because of a known issue downloading CVEs.

**Pull request merged**